src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -0,0 +1,517 @@
+        return;
+    }
+    uint32_t epochNow = std::chrono::duration_cast<chip::System::Clock::Seconds32>(cTMs).count();
+
+    if (forecast.Value().earliestStartTime.Value().IsNull())
+    {
+        /* Null means - We can start immediately */
+        earliestStartTime = epochNow; /* NOW */
+    }
+    else
+    {
+        earliestStartTime = forecast.Value().earliestStartTime.Value().Value();
+    }
+    /* Latest End Time is optional & cannot be null - unlike earliestStartTime! */
+    latestEndTime = forecast.Value().latestEndTime.Value();
+
+    duration = forecast.Value().endTime - forecast.Value().startTime; // the current entire forecast duration
+
+    if ((requestedStartTime < earliestStartTime) || ((requestedStartTime + duration) > latestEndTime))
+    {
+        ChipLogError(Zcl, "DEM: %s - Bad requestedStartTime %d.", __FUNCTION__, requestedStartTime);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+    else
+    {
+        ChipLogProgress(Zcl, "DEM: %s - Good requestedStartTime %d.", __FUNCTION__, requestedStartTime);
+
+        status = mDelegate.StartTimeAdjustRequest(requestedStartTime);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        if (status != Status::Success)
+        {
+            ChipLogError(Zcl, "DEM: %s - StartTimeAdjustRequest(%d) FAILURE", __FUNCTION__, requestedStartTime);
+            return;
+        }
+    }
+
+    return;
+}
+
+void Instance::HandlePauseRequest(HandlerContext & ctx, const Commands::PauseRequest::DecodableType & commandData)
+{
+    Status status                                               = Status::Success;
+    CHIP_ERROR err                                              = CHIP_NO_ERROR;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    uint32_t duration = commandData.duration;
+
+    if (forecast.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: %s - Forecast is Null", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    if (ESAStateEnum::kUserOptOut == mDelegate.GetESAState())
+    {
+        ChipLogError(Zcl, "DEM: %s - ESAState = kUserOptOut", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* value SHALL be between the MinPauseDuration and MaxPauseDuration indicated in the
+       ActiveSlotNumber index in the Slots list in the Forecast.
+    */
+    uint16_t activeSlotNumber;
+    if (forecast.Value().activeSlotNumber.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: %s - activeSlotNumber Is Null", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    activeSlotNumber = forecast.Value().activeSlotNumber.Value();
+    if (activeSlotNumber >= forecast.Value().slots.size())
+    {
```

Comment:
Shouldn't this actually be "impossible" ?
If so, an appropriate ...impossible... log msg ?

Reply:
Think the error message is ok now.